src/lib/integration_test_lib/malleable_error.ml

```diff
@@ -0,0 +1,351 @@
+
+let hard_error_string message =
+  Deferred.return
+    (Error
+       { Hard_fail.hard_error=
+           Test_error.raw_internal_error (Error.of_string message)
+       ; soft_errors= [] })
+
+let hard_errorf format = Printf.ksprintf hard_error_string format
+
+let soft_error res err =
+  Deferred.return
+    (Ok
+       { Accumulator.computation_result= res
+       ; soft_errors= [Test_error.raw_internal_error err] })
+
+let hard_error err =
+  Deferred.return
+    (Error
+       { Hard_fail.hard_error= Test_error.raw_internal_error err
+       ; soft_errors= [] })
+
+let hard_error_to_string (err : 'a t) : string Deferred.t =
+  let open Deferred.Let_syntax in
+  match%map err with
+  | Ok {Accumulator.computation_result= _; soft_errors= _} ->
+      "<none>"
+  | Error {Hard_fail.hard_error= hard_err; Hard_fail.soft_errors= _} ->
+      Error.to_string_hum hard_err.error
+
+let of_or_error_hard (or_err : 'a Or_error.t) : 'a t =
+  match or_err with
+  | Ok x ->
+      Deferred.return (Ok {Accumulator.computation_result= x; soft_errors= []})
+  | Error err ->
+      Deferred.return
+        (Error
+           { Hard_fail.hard_error= Test_error.raw_internal_error err
+           ; soft_errors= [] })
+
+let of_or_error_soft (res : 'a) (or_err : 'a Or_error.t) : 'a t =
+  match or_err with
+  | Ok x ->
+      Deferred.return (Ok {Accumulator.computation_result= x; soft_errors= []})
+  | Error err ->
+      Deferred.return
+        (Ok
+           { Accumulator.computation_result= res
+           ; soft_errors= [Test_error.raw_internal_error err] })
+
+let combine_errors (malleable_errors : 'a t list) : 'a list t =
+  let open T.Let_syntax in
+  let%map values =
+    List.fold_left malleable_errors ~init:(return []) ~f:(fun acc el ->
+        let%bind t = acc in
+        let%map h = el in
+        h :: t )
+  in
+  List.rev values
+
+let try_with (type a) ?(backtrace = false) (f : unit -> a) : a t =
+  of_or_error_hard (Or_error.try_with ~backtrace f)
+
+let of_option opt msg : 'a t =
+  Option.value_map opt
+    ~default:
+      (Deferred.return
+         (Error
+            { Hard_fail.hard_error=
+                Test_error.raw_internal_error (Error.of_string msg)
+            ; Hard_fail.soft_errors= [] }))
+    ~f:T.return
+
+let lift_error_set (type a) (m : a t) :
+    (a * Test_error.Set.t, Test_error.Set.t) Deferred.Result.t =
+  let open Deferred.Let_syntax in
+  let internal_error err = Test_error.Internal_error err in
+  let internal_error_set hard_errors soft_errors =
+    { Test_error.Set.hard_errors= List.map hard_errors ~f:internal_error
+    ; soft_errors= List.map soft_errors ~f:internal_error }
+  in
+  match%map m with
+  | Ok {Accumulator.computation_result; soft_errors} ->
+      Ok (computation_result, internal_error_set [] soft_errors)
+  | Error {Hard_fail.hard_error; soft_errors} ->
+      Error (internal_error_set [hard_error] soft_errors)
+
+module Map = struct
+  let rec malleable_error_list_iter ls ~f =
```

Comment:
The correct name of this value, for this pattern, would be merely `iter`. Similar for the other functions in this module.